prometheus/promsafe/safe.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package promsafe provides a layer of type-safety for label management in Prometheus metrics.
+//
+// Promsafe introduces type-safe labels, ensuring that the labels used with
+// Prometheus metrics are explicitly defined and validated at compile-time. This
+// eliminates common runtime errors caused by mislabeling, such as typos or
+// incorrect label orders.
+//
+// The following example demonstrates how to create and use a CounterVec with
+// type-safe labels (compared to how it's done in a regular way):
+//
+//	package main
+//
+//	import (
+//		"strconv"
+//
+//		"github.com/prometheus/client_golang/prometheus"
+//		"github.com/prometheus/client_golang/prometheus/promsafe"
+//	)
+//
+//	// Original unsafe way (no type safety)
+//
+//	func originalUnsafeWay() {
+//		counterVec := prometheus.NewCounterVec(
+//			prometheus.CounterOpts{
+//				Name: "http_requests_total",
+//				Help: "Total number of HTTP requests by status code and method.",
+//			},
+//			[]string{"code", "method"}, // Labels defined as raw strings
+//		)
+//
+//		// No compile-time checks; label order and types must be correct
+//		// You have to know which and how many labels are expected (in proper order)
+//		counterVec.WithLabelValues("200", "GET").Inc()
+//
+//		// or you can use map, that is even more fragile
+//		counterVec.With(prometheus.Labels{"code": "200", "method": "GET"}).Inc()
+//	}
+//
+//	// Safe way (Quick implementation, reflect-based under-the-hood)
+//
+//	type Labels1 struct {
+//		promsafe.StructLabelProvider
+//		Code   int
+//		Method string
+//	}
+//
+//	func safeReflectWay() {
+//		counterVec := promsafe.NewCounterVec[Labels1](prometheus.CounterOpts{
+//			Name: "http_requests_total_reflection",
+//			Help: "Total number of HTTP requests by status code and method (reflection-based).",
+//		})
+//
+//		// Compile-time safe and readable; Will be converted into properly ordered list: "200", "GET"
+//		counterVec.With(Labels1{Method: "GET", Code: 200}).Inc()
+//	}
+//
+//	// Safe way with manual implementation (no reflection overhead, as fast as original)
+//
+//	type Labels2 struct {
+//		promsafe.StructLabelProvider
+//		Code   int
+//		Method string
+//	}
+//
+//	func (c Labels2) ToPrometheusLabels() prometheus.Labels {
```

Comment:
Here one could implement constraints as well perhaps to contain values (cardinality). We have some "hidden" logic for this [here](https://pkg.go.dev/github.com/prometheus/client_golang@v1.20.5/prometheus#ConstrainableLabels)